File: Minimal/tests/Brontide.Minimal.PersistentInformation.Tests/PersistentInformationTests.fs
namespace Brontide.Minimal.PersistentInformation.Tests

open System
open NUnit.Framework
open Brontide.Minimal.Model
open Brontide.Minimal.Kernel
open Brontide.Minimal.Experimental.PersistentInformation

[<TestFixture>]
type PersistentInformationTests() =
    let get = function Ok value -> value | Error error -> failwith error.Code
    let error = function Error value -> value | Ok _ -> failwith "Expected a refusal."
    let name value = CanonicalName.create value
    let role = StoreRoleId.create "core"
    let corpus () =
        OpaqueCorpus.create
            (CorpusId.create "settings")
            "1"
            (Some SingleWriter)
            [ { Id = role; IdentityBearing = true; Required = true; AbsenceBehavior = DatasetUnavailable } ]
        |> get
    let actor () =
        let timeDomain = TimeDomainReference.create (name "Brontide.Minimal.Tests:FixtureTime")
        let initial = World.create (Guid.NewGuid()) timeDomain
        let created, _ =
            World.genesis
                (name "Brontide.Minimal.Tests:FixturePolicy")
                { Milliseconds = 0L; TimeDomain = timeDomain; UncertaintyMilliseconds = None }
                (fun genesis world -> Genesis.actor genesis (name "Brontide.Minimal.Tests:FixtureActor") world)
                initial
            |> Result.defaultWith failwith
        created.Reference

    [<Test>]
    member _.``C1 Corpus requires explicit supported concurrency and identity role`` () =
        let definition = { Id = role; IdentityBearing = true; Required = true; AbsenceBehavior = DatasetUnavailable }
        let missing = OpaqueCorpus.create (CorpusId.create "settings") "1" None [ definition ]
        let unsupported = OpaqueCorpus.create (CorpusId.create "settings") "1" (Some ExternalCoordination) [ definition ]
        let noIdentity = OpaqueCorpus.create (CorpusId.create "settings") "1" (Some SingleWriter) [ { definition with IdentityBearing = false } ]
        Assert.That((error missing).Code, Is.EqualTo "corpus-invalid")
        Assert.That((error unsupported).Code, Is.EqualTo "concurrency-unsupported")
        Assert.That((error noIdentity).Code, Is.EqualTo "corpus-invalid")

    [<Test>]
    member _.``C2 Dataset issuance is an authorized effect and denials are silent`` () =
        let registry = DatasetRegistry()
        let store = MemoryStore(StoreId.create "primary", Set.singleton Durable)
        let timeDomain = TimeDomainReference.create (name "Brontide.Minimal.Tests:PersistentTime")
        let mutable issuer = Unchecked.defaultof<Actor>
        let mutable stranger = Unchecked.defaultof<Actor>
        let mutable target = Unchecked.defaultof<Actor>
        let mutable otherTarget = Unchecked.defaultof<Actor>
        let mutable grant = Unchecked.defaultof<Capability>
        let mutable wrongTargetGrant = Unchecked.defaultof<Capability>
        let createOperation: OperationReference = { Name = name "Brontide.Minimal.Tests:Dataset.Create" }
        let appendOperation: OperationReference = { Name = name "Brontide.Minimal.Tests:Dataset.Append" }
        let otherOperation: OperationReference = { Name = name "Brontide.Minimal.Tests:Other.Create" }
        let initial = World.create (Guid.NewGuid()) timeDomain
        let _, world =
            World.genesis (name "Brontide.Minimal.Tests:PersistentBootstrap")
                { Milliseconds = 0L; TimeDomain = timeDomain; UncertaintyMilliseconds = None }
                (fun genesis world ->
                    let issued, world = Genesis.actor genesis (name "Brontide.Minimal.Tests:Issuer") world
                    issuer <- issued
                    let other, world = Genesis.actor genesis (name "Brontide.Minimal.Tests:Stranger") world
                    stranger <- other
                    let service, world = Genesis.actor genesis (name "Brontide.Minimal.Tests:DatasetService") world
                    target <- service
                    let otherService, world = Genesis.actor genesis (name "Brontide.Minimal.Tests:OtherService") world
                    otherTarget <- otherService
                    let operation = { Reference = createOperation; Description = "issue Dataset"; Target = target.Reference; CommandShape = BuiltIn.unitShape; ResultShape = BuiltIn.unitShape; Constraints = [] }
                    let world = World.registerOperation operation world |> Result.defaultWith failwith
                    let appendDefinition = { Reference = appendOperation; Description = "append Dataset"; Target = target.Reference; CommandShape = BuiltIn.unitShape; ResultShape = BuiltIn.unitShape; Constraints = [] }
                    let world = World.registerOperation appendDefinition world |> Result.defaultWith failwith
                    let otherDefinition = { Reference = otherOperation; Description = "unrelated"; Target = otherTarget.Reference; CommandShape = BuiltIn.unitShape; ResultShape = BuiltIn.unitShape; Constraints = [] }
                    let world = World.registerOperation otherDefinition world |> Result.defaultWith failwith
                    let issuedGrant, world = Genesis.capability genesis (name "Brontide.Minimal.Tests:DatasetGrant") issuer.Reference target.Reference (Set.ofList [ createOperation; appendOperation ]) [] false world |> Result.defaultWith failwith
                    grant <- issuedGrant
                    let otherGrant, world = Genesis.capability genesis (name "Brontide.Minimal.Tests:WrongTargetGrant") issuer.Reference otherTarget.Reference (Set.singleton otherOperation) [] false world |> Result.defaultWith failwith
                    wrongTargetGrant <- otherGrant
                    (), world)
                initial
            |> Result.defaultWith failwith
        let request actor targetActor capability operation : ExecutionRequest = { Initiator = actor; Target = targetActor; PresentedCapability = capability; Operation = operation; Command = UnitValue; Occurrence = None; Context = Map.empty }
        let mutable effects = 0
        let handler (request: ExecutionRequest) =
            effects <- effects + 1
            if request.Operation = createOperation then
                registry.Issue({ Issuer = request.Initiator; IssuingOperation = request.Operation }, corpus(), DatasetId.create "dataset-1", Map.ofList [ role, store :> IStoreEndpoint ]) |> ignore
            else
                registry.Append(DatasetId.create "dataset-1", role, SingleWriter, "value") |> ignore
            Ok(UnitValue, [], [])
        let environment = { TrustedTime = { Milliseconds = 1L; TimeDomain = timeDomain; UncertaintyMilliseconds = None }; ConstraintEvaluators = Map.empty; Handlers = Map.ofList [ createOperation, handler; appendOperation, handler ] }
        let deniedActor = World.step environment world (request stranger.Reference target.Reference grant.Reference createOperation)
        let deniedTarget = World.step environment world (request issuer.Reference target.Reference wrongTargetGrant.Reference createOperation)
        let deniedOperation = World.step environment world (request issuer.Reference target.Reference grant.Reference otherOperation)
        Assert.That(deniedActor.Outcome.Status, Is.EqualTo Denied)
        Assert.That(deniedTarget.Outcome.Status, Is.EqualTo Denied)
        Assert.That(deniedOperation.Outcome.Status, Is.EqualTo Denied)
        Assert.That(effects, Is.Zero)
        Assert.That(registry.Datasets, Is.Empty)
        let accepted = World.step environment world (request issuer.Reference target.Reference grant.Reference createOperation)
        Assert.That(accepted.Outcome.Status, Is.EqualTo Succeeded)
        Assert.That(registry.Datasets.Head.Issuer, Is.EqualTo issuer.Reference)
        let deniedAppend = World.step environment accepted.World (request stranger.Reference target.Reference grant.Reference appendOperation)
        Assert.That(deniedAppend.Outcome.Status, Is.EqualTo Denied)
        Assert.That(store.AppendCount, Is.Zero)
        let acceptedAppend = World.step environment accepted.World (request issuer.Reference target.Reference grant.Reference appendOperation)
        Assert.That(acceptedAppend.Outcome.Status, Is.EqualTo Succeeded)
        Assert.That(store.AppendCount, Is.EqualTo 1)

    [<Test>]
    member _.``C3 Dataset identity survives Store content loss`` () =
        let registry = DatasetRegistry()
        let store = MemoryStore(StoreId.create "primary", Set.singleton Durable)
        let id = DatasetId.create "dataset-identity"
        let createOperation: OperationReference = { Name = name "Brontide.Minimal.Tests:Create" }
        registry.Issue({ Issuer = actor(); IssuingOperation = createOperation }, corpus(), id, Map.ofList [ role, store :> IStoreEndpoint ]) |> ignore
        registry.Append(id, role, SingleWriter, "value") |> ignore
        store.Clear()
        Assert.That(registry.Datasets.Head.Id, Is.EqualTo id)
        Assert.That((registry.Read(id, role, SingleWriter) |> get), Is.Empty)

    [<Test>]
    member _.``C4 Dataset operations fail before Store effects at role and concurrency boundaries`` () =
        let registry = DatasetRegistry()
        let store = MemoryStore(StoreId.create "primary", Set.singleton Durable)
        let id = DatasetId.create "dataset-boundaries"
        let createOperation: OperationReference = { Name = name "Brontide.Minimal.Tests:Create" }
        registry.Issue({ Issuer = actor(); IssuingOperation = createOperation }, corpus(), id, Map.ofList [ role, store :> IStoreEndpoint ]) |> ignore
        let wrongRole = registry.Append(id, StoreRoleId.create "unknown", SingleWriter, "x")
        let wrongMode = registry.Append(id, role, ExternalCoordination, "x")
        Assert.That((error wrongRole).Code, Is.EqualTo "role-not-found")
        Assert.That((error wrongMode).Code, Is.EqualTo "concurrency-mismatch")
        Assert.That(store.AppendCount, Is.Zero)

    [<Test>]
    member _.``C5 Router guarantees are declared stable and do not leak backing guarantees`` () =
        let first = MemoryStore(StoreId.create "first", Set.ofList [ Durable; Encrypted ])
        let second = MemoryStore(StoreId.create "second", Set.singleton Durable)
        let router = RouterEndpoint.create (RouterId.create "router") (Set.singleton Durable) [ first; second ] false |> get
        Assert.That(router.Guarantees = Set.singleton Durable, Is.True)
        Assert.That(router.Select(second.Id) |> Result.isOk, Is.True)
        Assert.That(router.Guarantees = Set.singleton Durable, Is.True)

    [<Test>]
    member _.``C6 Router fallback refusal and topology redaction are explicit`` () =
        let first = MemoryStore(StoreId.create "first", Set.singleton Durable)
        first.IsAvailable <- false
        let second = MemoryStore(StoreId.create "second", Set.singleton Durable)
        let router = RouterEndpoint.create (RouterId.create "router") (Set.singleton Durable) [ first; second ] false |> get
        Assert.That(router.Append "fallback" |> Result.isOk, Is.True)
        Assert.That(second.Read() = [ "fallback" ], Is.True)
        Assert.That((router.Describe false).SelectedBacking.IsNone, Is.True)
        Assert.That((router.Describe true).SelectedBacking.IsNone, Is.True)
        let inspectable = RouterEndpoint.create (RouterId.create "inspectable") (Set.singleton Durable) [ second ] true |> get
        Assert.That((inspectable.Describe false).SelectedBacking.IsNone, Is.True)
        Assert.That((inspectable.Describe true).SelectedBacking, Is.EqualTo(Some second.Id))
        let unsupported = RouterEndpoint.create (RouterId.create "bad") (Set.singleton Encrypted) [ second ] true
        Assert.That((error unsupported).Code, Is.EqualTo "router-guarantee-unsupported")

    [<Test>]
    member _.``C7 identity spaces remain distinct public value types`` () =
        Assert.That(typeof<CorpusId>, Is.Not.EqualTo typeof<DatasetId>)
        Assert.That(typeof<StoreId>, Is.Not.EqualTo typeof<RouterId>)

    [<Test>]
    member _.``C8 every failed operation preserves Store observations`` () =
        let store = MemoryStore(StoreId.create "store", Set.singleton Durable)
        let registry = DatasetRegistry()
        Assert.That(registry.Append(DatasetId.create "missing", role, SingleWriter, "x") |> Result.isError, Is.True)
        Assert.That(store.AppendCount, Is.Zero)

File: Reference/tests/Brontide.Reference.PersistentInformation.Tests/PersistentInformationTests.cs
using Brontide.Reference.Core;
using Brontide.Reference.Experimental.PersistentInformation;
using NUnit.Framework;

namespace Brontide.Reference.PersistentInformation.Tests;

public sealed class PersistentInformationTests
{
    private static readonly StoreRoleId CoreRole = StoreRoleId.Parse("core");

    [Test]
    public void C1_Corpus_requires_explicit_supported_concurrency_and_identity_role()
    {
        var role = new StoreRoleDefinition(CoreRole, true, true, StoreRoleAbsenceBehavior.DatasetUnavailable);

        Assert.That(OpaqueCorpus.Create(CorpusId.Parse("settings"), "1", null, [role]).Code,
            Is.EqualTo("corpus-invalid"));
        Assert.That(OpaqueCorpus.Create(CorpusId.Parse("settings"), "1", ConcurrentAccessMode.ExternalCoordination, [role]).Code,
            Is.EqualTo("concurrency-unsupported"));
        Assert.That(OpaqueCorpus.Create(CorpusId.Parse("settings"), "1", ConcurrentAccessMode.SingleWriter,
            [role with { IdentityBearing = false }]).Code, Is.EqualTo("corpus-invalid"));
    }

    [Test]
    public async Task C2_Dataset_issuance_is_an_authorized_effect_and_denials_are_silent()
    {
        var registry = new DatasetRegistry();
        var store = new InMemoryStore(StoreId.Parse("primary"), EndpointGuarantee.Durable);
        var corpus = Corpus();
        ActorReference issuer = null!;
        ActorReference stranger = null!;
        ActorReference target = null!;
        ActorReference otherTarget = null!;
        Capability createGrant = null!;
        Capability wrongTargetGrant = null!;
        var create = OperationReference.Parse("Dataset.Create");
        var append = OperationReference.Parse("Dataset.Append");
        var other = OperationReference.Parse("Other.Create");

        var domain = AuthorityDomain.Create("persistent-information", genesis =>
        {
            issuer = genesis.Actor("Issuer");
            stranger = genesis.Actor("Stranger");
            target = genesis.Actor("Dataset service");
            otherTarget = genesis.Actor("Other service");
            genesis.Operation(create, target, ShapeContract.Unit, ShapeContract.Unit, "issue Dataset", context =>
            {
                var issued = registry.Issue(
                    new DatasetIssuance(context.Execution.Initiator, context.Execution.Operation),
                    corpus,
                    DatasetId.Parse("dataset-1"),
                    new Dictionary<StoreRoleId, IStoreEndpoint> { [CoreRole] = store });
                if (!issued.IsSuccess)
                {
                    throw new InvalidOperationException(issued.Failure!.Reason);
                }

                return OperationEffect.SucceededAsync(ShapeValue.Unit);
            });
            genesis.Operation(append, target, ShapeContract.Unit, ShapeContract.Unit, "append Dataset", _ =>
            {
                var appended = registry.Append(
                    DatasetId.Parse("dataset-1"), CoreRole, ConcurrentAccessMode.SingleWriter, "value");
                if (!appended.IsSuccess)
                {
                    throw new InvalidOperationException(appended.Failure!.Reason);
                }

                return OperationEffect.SucceededAsync(ShapeValue.Unit);
            });
            genesis.Operation(other, otherTarget, ShapeContract.Unit, ShapeContract.Unit, "unrelated",
                _ => OperationEffect.SucceededAsync(ShapeValue.Unit));
            createGrant = genesis.Grant(issuer, target, [create, append]);
            wrongTargetGrant = genesis.Grant(issuer, otherTarget, [other]);
        });

        var wrongActor = await domain.ExecuteAsync(stranger, create, createGrant, ShapeValue.Unit);
        var wrongTarget = await domain.ExecuteAsync(issuer, create, wrongTargetGrant, ShapeValue.Unit);
        var wrongOperation = await domain.ExecuteAsync(issuer, other, createGrant, ShapeValue.Unit);
        Assert.That(wrongActor.Outcome.Status, Is.EqualTo(OutcomeStatus.Rejected));
        Assert.That(wrongTarget.Outcome.Status, Is.EqualTo(OutcomeStatus.Rejected));
        Assert.That(wrongOperation.Outcome.Status, Is.EqualTo(OutcomeStatus.Rejected));
        Assert.That(registry.Datasets, Is.Empty);
        Assert.That(store.AppendCount, Is.Zero);

        var accepted = await domain.ExecuteAsync(issuer, create, createGrant, ShapeValue.Unit);
        Assert.That(accepted.Outcome.Status, Is.EqualTo(OutcomeStatus.Succeeded));
        Assert.That(registry.Datasets.Single().Issuer, Is.SameAs(issuer));
        Assert.That(registry.Datasets.Single().IssuingOperation, Is.EqualTo(create));

        var deniedAppend = await domain.ExecuteAsync(stranger, append, createGrant, ShapeValue.Unit);
        Assert.That(deniedAppend.Outcome.Status, Is.EqualTo(OutcomeStatus.Rejected));
        Assert.That(store.AppendCount, Is.Zero);
        var acceptedAppend = await domain.ExecuteAsync(issuer, append, createGrant, ShapeValue.Unit);
        Assert.That(acceptedAppend.Outcome.Status, Is.EqualTo(OutcomeStatus.Succeeded));
        Assert.That(store.AppendCount, Is.EqualTo(1));
    }

    [Test]
    public void C3_Dataset_identity_survives_store_content_loss()
    {
        var registry = new DatasetRegistry();
        var store = new InMemoryStore(StoreId.Parse("primary"), EndpointGuarantee.Durable);
        var datasetId = DatasetId.Parse("dataset-identity");
        var issuance = new DatasetIssuance(Actor(), OperationReference.Parse("Dataset.Create"));
        Assert.That(registry.Issue(issuance, Corpus(), datasetId,
            new Dictionary<StoreRoleId, IStoreEndpoint> { [CoreRole] = store }).IsSuccess, Is.True);
        Assert.That(registry.Append(datasetId, CoreRole, ConcurrentAccessMode.SingleWriter, "value").IsSuccess, Is.True);

        store.Clear();

        Assert.That(registry.Datasets.Single().Id, Is.EqualTo(datasetId));
        Assert.That(registry.Read(datasetId, CoreRole, ConcurrentAccessMode.SingleWriter).Value, Is.Empty);
    }

    [Test]
    public void C4_Dataset_operations_fail_before_store_effects_at_role_and_concurrency_boundaries()
    {
        var registry = new DatasetRegistry();
        var store = new InMemoryStore(StoreId.Parse("primary"), EndpointGuarantee.Durable);
        var datasetId = DatasetId.Parse("dataset-boundaries");
        registry.Issue(new DatasetIssuance(Actor(), OperationReference.Parse("Dataset.Create")), Corpus(), datasetId,
            new Dictionary<StoreRoleId, IStoreEndpoint> { [CoreRole] = store });

        Assert.That(registry.Append(datasetId, StoreRoleId.Parse("unknown"), ConcurrentAccessMode.SingleWriter, "x").Code,
            Is.EqualTo("role-not-found"));
        Assert.That(registry.Append(datasetId, CoreRole, ConcurrentAccessMode.ExternalCoordination, "x").Code,
            Is.EqualTo("concurrency-mismatch"));
        Assert.That(store.AppendCount, Is.Zero);
    }

    [Test]
    public void C5_Router_guarantees_are_declared_stable_and_do_not_leak_backing_guarantees()
    {
        var first = new InMemoryStore(StoreId.Parse("first"), EndpointGuarantee.Durable, EndpointGuarantee.Encrypted);
        var second = new InMemoryStore(StoreId.Parse("second"), EndpointGuarantee.Durable);
        var created = RouterEndpoint.Create(RouterId.Parse("router"), [EndpointGuarantee.Durable], [first, second], false);
        var router = created.Value!;

        Assert.That(router.Guarantees, Is.EquivalentTo(new[] { EndpointGuarantee.Durable }));
        Assert.That(router.Guarantees, Does.Not.Contain(EndpointGuarantee.Encrypted));
        Assert.That(router.Select(second.Id).IsSuccess, Is.True);
        Assert.That(router.Guarantees, Is.EquivalentTo(new[] { EndpointGuarantee.Durable }));
        Assert.That(router.Id, Is.EqualTo(RouterId.Parse("router")));
    }

    [Test]
    public void C6_Router_falls_back_refuses_unsupported_guarantees_and_redacts_topology()
    {
        var first = new InMemoryStore(StoreId.Parse("first"), EndpointGuarantee.Durable) { IsAvailable = false };
        var second = new InMemoryStore(StoreId.Parse("second"), EndpointGuarantee.Durable);
        var router = RouterEndpoint.Create(RouterId.Parse("router"), [EndpointGuarantee.Durable], [first, second], false).Value!;

        Assert.That(router.Append("fallback").IsSuccess, Is.True);
        Assert.That(second.Read(), Is.EqualTo(new[] { "fallback" }));
        Assert.That(router.Describe(false).SelectedBacking, Is.Null);
        Assert.That(router.Describe(true).SelectedBacking, Is.Null);
        var inspectable = RouterEndpoint.Create(RouterId.Parse("inspectable"), [EndpointGuarantee.Durable], [second], true).Value!;
        Assert.That(inspectable.Describe(false).SelectedBacking, Is.Null);
        Assert.That(inspectable.Describe(true).SelectedBacking, Is.EqualTo(second.Id));
        Assert.That(RouterEndpoint.Create(RouterId.Parse("bad"), [EndpointGuarantee.Encrypted], [second], true).Code,
            Is.EqualTo("router-guarantee-unsupported"));
    }

    [Test]
    public void C7_Identity_spaces_are_distinct_public_value_types()
    {
        Assert.Multiple(() =>
        {
            Assert.That(default(CorpusId), Is.Not.InstanceOf<DatasetId>());
            Assert.That(default(DatasetId), Is.Not.InstanceOf<StoreRoleId>());
            Assert.That(default(StoreId), Is.Not.InstanceOf<RouterId>());
        });
    }

    [Test]
    public void C8_All_failure_paths_preserve_store_observations()
    {
        var store = new InMemoryStore(StoreId.Parse("store"), EndpointGuarantee.Durable);
        var registry = new DatasetRegistry();
        var dataset = DatasetId.Parse("missing");

        Assert.That(registry.Append(dataset, CoreRole, ConcurrentAccessMode.SingleWriter, "x").IsSuccess, Is.False);
        Assert.That(store.AppendCount, Is.Zero);
    }

    private static OpaqueCorpus Corpus() => OpaqueCorpus.Create(
        CorpusId.Parse("settings"),
        "1",
        ConcurrentAccessMode.SingleWriter,
        [new StoreRoleDefinition(CoreRole, true, true, StoreRoleAbsenceBehavior.DatasetUnavailable)]).Value!;

    private static ActorReference Actor()
    {
        ActorReference actor = null!;
        _ = AuthorityDomain.Create("actor-fixture", genesis => actor = genesis.Actor("Issuer"));
        return actor;
    }
}
